packages/venia-ui/lib/queries/getCategory.graphql

```diff
@@ -29,17 +29,64 @@ query category(
             }
         }
         items {
+            # Once graphql-ce/1027 is resolved, use a ProductDetails fragment here instead.
```

Comment:
I have concerns about requesting the entire product details fragment here in this request. This would be a form of "pre-fetching" wherein we would fetch _all_ the product details for the visible products on the category page before a user has navigated to the products. I don't think we should do this unless we can make a strong argument for it.

I'll detail the pros/cons. Please feel free to add!

Pros:

* Any product a user clicks on from the category view will have been cached so it should render faster.
* This query is more composable in that `items` will just use the Fragment. 

Cons:

* More data is fetched on the category view so the category view may render more slowly.
* Extra data is fetched that isn't ever used on the category page.
* If we remove something the category requires for rendering from the pdp fragment we may introduce a bug without realizing it.

Reply:
Regarding cons 1 and 2, we can have the SW do this for us. It can request in the back, so no lag/load on the main thread and we can write logic into it to only request if the client is running on WiFi.

Reply:
Thanks for bringing this up! Yes, this is the tradeoff.

I made some inline edits to your pros & cons 😄 , we'll definitely be over-fetching on the Category page in order to make the PDP load faster.

Fortunately, this is gzipped JSON (string data) so - at least currently - we're talking about adding an extra ~2.5KB. The last `graphql` request shown is the one that requests category data:

This PR:

![Screen Shot 2019-10-25 at 8 16 22 AM](https://user-images.githubusercontent.com/13182778/67574376-ee807480-f6ff-11e9-9b12-88532c705a5a.png)

Develop (upward-js):

![Screen Shot 2019-10-25 at 8 17 07 AM](https://user-images.githubusercontent.com/13182778/67574416-05bf6200-f700-11e9-8adc-549d10bfc632.png)

To your point though, the timing does get blown out a bit. I'd say that one is my local machine vs. a deployed instance but all the other requests took about the same time so this may in fact be an issue.

> * If we remove something the category requires for rendering from the pdp fragment we may introduce a bug without realizing it.

This is a good point and yeah we'll have to keep in mind that this fragment is definitely not PDP-specific.

Reply:
There is a "middle ground" that I should point out too: the idea of "partial data". From [the useQuery docs](https://www.apollographql.com/docs/react/data/queries/#usequery-api):

> returnPartialData: Opt into receiving partial results from the cache for queries that are not fully satisfied by the cache. false by default.

Because this is `false` currently, we have to ensure that all the fields of a product are present in the cache for a cache hit to take place and avoid going to the network. This is why the category query adds all of the product fields.

If we change this to `true`, we could start rendering the PDP with whatever data we have in the cache for that product while we query the network for the rest. Currently the Category page requests `name`, `price`, image, etc. This will lead to the PDP initially rendering with only that data. When the server response comes back, then we'd fill in the rest of the details.

This would obviously take some refactoring of the PDP to accomplish as well. Mainly, it wouldn't be able to trust that all of its data will be present.

Reply:
> If we change this to true, we could start rendering the PDP with whatever data we have in the cache for that product while we query the network for the rest.

This is *exactly* what the upcoming [React Concurrent Mode](https://reactjs.org/docs/concurrent-mode-suspense.html) enables and prescribes. They call this strategy `render-as-you-fetch`, as opposed to `fetch-on-render` or `fetch-then-render`.

> This would obviously take some refactoring of the PDP to accomplish as well. Mainly, it wouldn't be able to trust that all of its data will be present.

And since React will soon provide us with tools to do this canonically, I'll recommend that we hold off on restructuring things until then. Should come together nicely very soon.

Reply:
> I have concerns about requesting the entire product details fragment here in this request.

As you noted, this is just another form of prefetching, and as Revanth & Andy noted, it's a small amount of extra bytes that we can request in the background. Given how slow GraphQL calls are right now, a cache hit when a product is clicked would probably be well worth the potential overfetch.

Reply:
> Given how slow GraphQL response times are right now, a cache hit when a product is clicked would probably be well worth the potential overfetch.

I still have concerns but I guess this would be a pro :)

Reply:
Yeah, I'm in favor of this. I understand and agree with @sirugh 's concerns as things to watch. Thank you (and you Andy) for enumerating them so clearly. Still, this is exactly the right use case for a fragment. Especially because Apollo understands fragments discretely, and can cache them by concrete type. So this should be really good.

Maybe we can get better opt in the future by splitting this fragment into a few smaller fragments that you can optionally compose together, but for now I don't think it's a problem. Partial fetch and React concurrent render should cover for the additional fetching, like @jimbo and @supernova-at say.